test(calendar): Add tests for calendar construction

mhchen opened the pull request:
Add some tests for calendar elements displayed upon initialization.
These tests are mostly focused around making sure a single month
displays properly.

mhchen commented:
@jelbourn please review at your convenience.

jelbourn commented on src/components/calendar/calendar.spec.js:
```diff
@@ -38,13 +40,106 @@ describe('md-checkbox', function() {
   describe('calendar construction', function() {
 
+    /**
+     * Extracts an text as an array from a tr element.
```
`Extracts text as an array (one element per cell) from a tr element.`

jelbourn commented on src/components/calendar/calendar.spec.js:
```diff
@@ -38,13 +40,106 @@ describe('md-checkbox', function() {
+    it('should display the weeks header in the first row', function() {
+      var header = element.querySelector('.md-calendar-day-header tr');
+
+      expect(extractRowText(header)).toEqual(['S', 'M', 'T', 'W', 'T', 'F' ,'S']);
```
This is okay for now, but eventually it would be better to have a test that makes sure the values come from $mdDateLocale

jelbourn commented on src/components/calendar/calendar.spec.js:
```diff
@@ -38,13 +40,106 @@ describe('md-checkbox', function() {
+    });
+
+    it('should render the displayed month correctly in a table', function() {
+      pageScope.myDate = new Date(2014, 4, 30);
```
Since more `new Date` calls are happening now, can you copy how I'm doing the month indices in `dateUtil.spec.js` ?

jelbourn commented on src/components/calendar/calendar.spec.js:
```diff
@@ -38,13 +40,106 @@ describe('md-checkbox', function() {
+          return td;
+        }
+      }
+    }
```
Can you move these helper functions to the top of the file (under `applyDateChange`) ?

jelbourn commented on src/components/calendar/calendar.spec.js:
```diff
@@ -38,13 +40,106 @@ describe('md-checkbox', function() {
+     * Extracts an text as an array from a tr element.
+     */
+    function extractRowText(tr) {
+      var headerTexts = [];
```
`headerTexts` > `cellContents`

jelbourn commented on src/components/calendar/calendar.spec.js:
```diff
@@ -38,13 +40,106 @@ describe('md-checkbox', function() {
+      pageScope.myDate = new Date(2014, 4, 30);
+      applyDateChange();
+
+      var calendarRows = element.querySelectorAll('.md-calendar-month:first-child tr');
```
This will break once the changes for infinite scrolling are in; there will be more months in either direction rendered at one time.

jelbourn commented on src/components/calendar/calendar.spec.js:
```diff
@@ -38,13 +40,106 @@ describe('md-checkbox', function() {
+      expect(calendarDates).toEqual(expectedDates);
+    });
+
+    it('should show only the month on the first row if the 1st of the month is before Tuesday', function() {
```
Keep lines under 100 characters

jelbourn commented on src/components/calendar/calendar.spec.js:
```diff
@@ -38,13 +40,106 @@ describe('md-checkbox', function() {
+      var tds = monthElement.querySelectorAll('td');
+      var td;
+
+      for (var i = 0; i < tds.length; i++) {
```
Could do this more quickly / concisely using the dateId

mhchen replied:
@jelbourn The reason I went with this somewhat-obscure approach is because the dateId is not exposed and is an implementation detail. Relying on it might break in case the dateId mechanism changes in the future. In addition the directiveId isn't exposed either, so I would assume 0 for the purposes of this test (which might also lead to brittleness). Just wanted to confirm my understanding and that these are acceptable tradeoffs.

jelbourn replied:
That's fair; I was thinking that it would be fine if the getDateId function was made public. Either way is fine.

jelbourn commented:
@mhchen comments made!

mhchen commented:
@jelbourn Pushed an amended commit to address your comments, and added some minor refactors (e.g. moved directive creation and compilation to a function)

jelbourn commented:
@mhchen the diff seems to be messed up now (showing entire calendar as new); can you rebase on `upstream wip/calendar` and re-push?

mhchen commented:
@jelbourn Done

jelbourn commented:
Closed via 133e60008167f319e6ddf4bdc723f099293618ee
